### DISCUSSION TITLE
States and plugins are badly designed?

### DISCUSSION THREAD

**Discussion by feelingsonice** [bevyengine/bevy#10950] [General]
The way that states and plugins have been designed in Bevy have me scratching my head. Although I'm quite new to Bevy so if there's something I'm misunderstanding please let me know.

In my mind, the point of states is so that systems/functions can be isolated into their corresponding states. Ofc, there's all sorts of state machine design patterns but the general idea stays the same. i.e. system `S` within state `A` can make certain assumptions about the current circumstance so that it can avoid the organizational hellholes.

Plugins, on the other hand, also kind of does this, but more from a code organization perspective. It isolates ECSs into their respective "chunk" of code.

All this is fine when your code is nice and simple, but when trying to create a more complexgame, there's going to be multiple plugins, states, and perhaps even nested states. It feels like the concept of "plugin" and "states" don't mix well.

To illustrate this, consider the `bevy_asset_loader` plugin, which uses states to separate out the "loading" state:

```rust
fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_state::<MyStates>()
        .add_loading_state(
            LoadingState::new(MyStates::Loading)
                .continue_to_state(MyStates::Main)
                .load_collection::<MyAssets>(),
        )
        .add_systems(OnEnter(MyStates::Main), my_setup)
        .add_systems(Update, my_system.run_if(in_state(AppState::Main)))
        .run();
}

#[derive(AssetCollection, Resource)]
struct MyAssets {
    #[asset(path = "some_audio.ogg")]
    background: Handle<AudioSource>,
}

#[derive(Clone, Eq, PartialEq, Debug, Hash, Default, States)]
enum MyStates {
    #[default]
    Loading,
    Main,
}
```

Notice how `my_setup` and `my_system` had to be manually defined under a state when added. Now, if I want to create a new plugin called `RunMyGamePlugin`, it now needs to be aware of the `MyStates` state (even though they have nothing to do with each other). If I want to add a system named `jump` to `RunMyGamePlugin`, it now has to import the `MyStates` state. If I want a `JumpAttack` nested state, I would need to first check it's in the `MyStates ::Main` state, and then the `Jump` state. You can see how this can get complicated if you have more plugins and need more nested states. 

Ideally, plugins and states are not aware of other parts of the app. Instead, the parent user is context aware and defines their relationships.

**Comment by alice-i-cecile** [comment]
See https://github.com/bevyengine/bevy/issues/2160 for previous thoughts along these lines.

**Reply by feelingsonice** [reply]
left a comment :)
